File: C_plus_plus/Find_Minimum_in_Rotated_Sorted_Array.cpp
class Solution {
public:
    int findMin(vector<int> &num) {
        int i = 0 , j = num.size()-1;
        if(num[0]<=num[j]) return num[0];
        while(i<=j){
            if((i+j)/2>i&&num[(i+j)/2]<num[(i+j)/2-1]) return num[(i+j)/2];
            if((i+j)/2<j&&num[(i+j)/2]>num[(i+j)/2+1]) return num[(i+j)/2+1];
            if(num[(i+j)/2]<num[i]) j = (i+j)/2-1;
            else i = (i+j)/2+1;
        }
    }
};

File: java/Find_Minimum_in_Rotated_Sorted_Array.java
// Binary Search
// If we retate the array, there will a sub array that is sorted, and the minimum one is in another subarray.
// Time is O(logn), space is O(1).

public class Solution {
    public int findMin(int[] nums) {
        int low = 0, high = nums.length-1;
        while (low < high) {
            int mid = (low+high)/2;
            if (nums[mid] < nums[low]) {
                high = mid;
            }
            else if (nums[mid] < nums[high]) {
                break;
            }
            else {
                low = mid+1;
            }
        }
        return nums[low];
    }
}
